File: src/main/scala/xieyuheng/cicada/Top.scala
package xieyuheng.cicada

sealed trait Top
final case class TopLet(name: String, exp: Exp) extends Top
final case class TopDefine(name: String, t: Exp, exp: Exp) extends Top
final case class TopKeywordRefuse(exp: Exp, t: Exp) extends Top
final case class TopKeywordAccept(exp: Exp, t: Exp) extends Top
final case class TopKeywordShow(exp: Exp) extends Top
final case class TopKeywordEq(rhs: Exp, lhs: Exp) extends Top

File: src/top.ts
import * as Exp from "./exp"

export abstract class Top {}

export class TopLet extends Top {
  constructor(
    public name: string,
    public exp: Exp.Exp,
  ) { super() }
}

export class TopDefine extends Top {
  constructor(
    public name: string,
    public t: Exp.Exp,
    public exp: Exp.Exp,
  ) { super() }
}

export class TopKeywordRefuse extends Top {
  constructor(
    public exp: Exp.Exp,
    public t: Exp.Exp,
  ) { super() }
}

export class TopKeywordAccept extends Top {
  constructor(
    public exp: Exp.Exp,
    public t: Exp.Exp,
  ) { super() }
}

export class TopKeywordShow extends Top {
  constructor(
    public exp: Exp.Exp,
  ) { super() }
}

export class TopKeywordEq extends Top {
  constructor(
    public rhs: Exp.Exp,
    public lhs: Exp.Exp,
  ) { super() }
}
